[1단계 - 자동차경주 구현] 키아라(김도희) 미션 제출합니다.

kiarakim opened the pull request:
리뷰 많이 해주세염..ㅎㅎ

kiarakim committed:
test : add StringTest

kiarakim committed:
docs : 기능 구현 목록 작성

kiarakim committed:
test : StringTest 작성

kiarakim committed:
feat : add skeleton code

kiarakim committed:
docs : 기능 구현 목록 갱신

kiarakim committed:
style : 코드 정렬

kiarakim committed:
feat : 자동차 이름 입력 받기

kiarakim committed:
docs : 시도 횟수 요청 문구 출력

kiarakim committed:
feat : 시도 횟수 요청 문구 출력

kiarakim committed:
docs : 자동차 전진 및 게임 진행

kiarakim committed:
feat : 자동차의 전진 및 보류 결정, 경주 진행하기

kiarakim committed:
docs : 자동차 최대 개수 지정 취소

kiarakim committed:
add setTest

kiarakim committed:
docs : 우승자 결정 및 출력 완료

kiarakim committed:
feat : 우승 결정 및 출력 기능

kiarakim committed:
refactor : depth 줄이기 및 매직넘버 제거

kiarakim committed:
docs : 자동차 이름이 같을 경우

kiarakim committed:
feat : 자동차 이름 중복 제거 기능 구현

kiarakim committed:
feat : 자동차 이름 중복 제거 기능 구현

kiarakim committed:
refactor : static 상수로 변경

kiarakim committed:
test : test 코드 작성

kiarakim committed:
refactor : Winner 클래스 test 위한 랜덤값 발생 위치 변경

sonypark commented on README.md:
```diff
@@ -5,3 +5,30 @@
 ## 우아한테크코스 코드리뷰
 
 - [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+
+- [o]  자동자 이름 요청 문구 출력
+- [o]  자동차 이름 입력 받기
+    - [o]  공백이 있다면 제거
+    - [o]  빈 값일 경우
+    - [o]  자동차 이름이 1이상 5이하인가?
+    - [o]  한 대만 있을 경우
+    - [x]  자동차 몇대까지 입력가능한지 지정
+    - [o]  쉼표 기준으로 자동차 분리
+    - [o]  자동차 이름이 같을 경우
+- [o]  시도 횟수 요청 문구 출력
+    - [o] 횟수 입력이 숫자가 아닌가?
+    - [o] 횟수 입력이 자연수가 아닌가?
+    - [o] 몇 번까지인지 지정
+    - [o] 빈 값일 경우
+    - [o] 0 이하일 경우
+- [o]  자동차의 전진 및 보류 결정
+    - [o]  한 trial마다 자동차 개수만큼 무작위 값 구함
+    - [o]  4이상일 경우 (거리)+1
+    - [o]  3이하일 경우 (거리)+0
+- [o]  경주 진행하기
+    - [o]  게임 진행하기
+    - [o]  각 trial별 실행 결과 보여주기
+- [o]  우승 결정
+    - [o]  제일 많이 전진한 거리 구하기
+    - [o]  제일 많이 전진한 자동차 찾기
+- [o]  결과 출력 - join
```
👍

sonypark commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,45 @@
+public class Controller {
+    private final InputView inputView = new InputView();
+    private final OutputView outputView = new OutputView();
+    private final Scanner scanner = new Scanner(System.in);
```
scanner 는 언제 필요한가요?
Controller가 가지고 있어야할까요?

kiarakim replied:
음, 아니요. 입력받는 곳에서 필요한 것 같아서 InputView에 넘겨주었습니다.

sonypark commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,47 @@
+public class Car implements Comparable<Car> {
+    private final static String ONE_TO_FIVE = "이름은 1이상 5이하입니다.";
+    private final String carName;
+    private int distance = 0;
```
코드 컨벤션을 확인해서 띄어쓰기 해주세요~

kiarakim replied:
네 감사합니다. 
혹시 static 변수 뒤에 빈줄이 있어야 하는건가요? 제가 띄어쓰기와 빈줄을 같다고 가정해보았는데요,
제가 아직 전역변수 간 띄어쓰기에 관한 코드 컨벤션은 찾지 못했습니다.
빈줄에 관한 내용으로

- import 선언 시 그룹끼르 묶고 그룹 사이엔 빈줄을 삽입한다
- 메서드 선언이 끝난 후 다음 메서드 선언이 시작되기 전 빈줄을 삽입한다.

등이 있었습니다.

그런데 코드를 살펴보던 중 첫 줄의 final과 static의 순서가 바뀌어 순서대로 둘 자리를 바꿔주었는데 혹시 이부분이 띄어쓰기와 관련된 피드백이었는지 궁금합니다!

sonypark replied:
네 자바 클래스내 변수,생성자,메서드의 위치에 대한 컨벤션이 있는데요. 여기서는 클래스 변수에 대한 선언 순서를 말씀드린거였습니다 :) 
static 변수를 상위에 두고 클래스 변수는 두고 그 사이는 한 줄 띄는 포맷을 사용합니다. 코드의 일관성과 가독성면에서 컨벤션을 지키는게 좋으니 앞으로 코드 작성하실 때 지켜주시면 좋을 것 같습니다!

- [Java coding convention about static method](https://stackoverflow.com/questions/7530535/java-coding-convention-about-static-method)
- [Class and Interface Declarations](https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.html#1852)
- [Code Examples](https://www.oracle.com/java/technologies/javase/codeconventions-codeexamples.html#182)

sonypark commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,47 @@
+    private void validateNameLength(String inputCar) {
+        if (inputCar.length() < 1 || inputCar.length() > 5) {
+            throw new IllegalArgumentException(ONE_TO_FIVE);
+        }
```
1, 5는 무엇을 의미하나요? 매직넘버를 줄여봅시다

kiarakim replied:
네 감사합니다. 구현 당시엔 매직넘버의 존재를 알았는데 리펙터링을 할 땐 잘 보이지 않아서 변경하지 못한 것 같습니다. 조금 더 주의를 기울여서 봐야할 것 같아요!

sonypark commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,47 @@
+    public void addDistance(int makeNumber) {
+        if (makeNumber >= 4) distance++;
+    }
+
```
- 여기서 4도 매직넘버네요
- 4이상이라는건 어떤걸 의미하나요?

kiarakim replied:
네. 4는 자동차가 움직일 수 있는 경계를 의미하고 있어서 문지방을 뜻하는 THRESHOLD 라고 상수 선언해주었습니다.

sonypark commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,47 @@
+    @Override
+    public int compareTo(Car car) {
+        return (this.distance > car.distance) ? 1 : -1;
+    }
```
자동차의 순위를 판단하는 로직이 Car에 있어야 할까요? 아니면 자동차 순위를 판단하는 쪽에 맡겨야 할까요?
자동차 순위에 대한 오름/내림 차순 기준이 바뀌게 되면 변경사항은 어디서 일어나야될까요?

kiarakim replied:
자동차의 순위를 판단하는 로직은 Car보단 Winner 클래스에 있는 것이 자연스러운 것 같습니다. 아무래도 Winner 클래스에서 최종 후보를 선별하여 우승자를 반환하기 때문입니다.
기존에 Winner 클래스에서 Collection.max()를 사용하기 위해 Car 클래스에서 Complarable<Car> 을 implement 해주고 compareTo() 메서드를 오버라이딩해준 것을 삭제한 후 Winner 클래스에서 순위 판별 로직 구현했습니다.
순위 판별을 위해
```
private Car findMaxDistance(List<Car> winners) {
	int distance = 0;
	for (Car car : winners) {
		if (distance < car.getDistance()) {
			distance = car.getDistance();
		}
	}
	return winners.get(0);
}
```
코드를 넣어주었고 추후에 오름/내림 차순의 기준이 바뀌게 될 경우 이곳에서 등호의 방향을 바꾸어 주는 방식으로 변경하면 될 것 같습니다.

sonypark replied:
네 역할에 맡게 로직을 옮긴것 좋습니다. 메서드를 만들때는 그 클래스의 책임이 맞는지 한번 더 생각해보면 도움이 될 것 같아요 👍

sonypark commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,65 @@
+    List<Car> cars = new ArrayList<>();
+    private final static String ERROR_NAME = "두 대 이상의 자동차 이름을 입력해주세요";
+    private final static String ERROR_DUPLICATION = "중복되지 않은 자동차를 입력해주세요";
+    private final static String COMMA = ",";
```
자바 코드 컨벤션을 지켜주세요~ 검색해서 공부해보시길 추천드립니다

kiarakim replied:
평소에 의식하고 쓰지 않은 결과인 것 같아요.
final과 static의 순서를 바꿔주었고,
앞으로 주의해서 코드 컨벤션 지키겠습니다. 감사합니다.

sonypark replied:
네 위에 코멘트 드린 내용 참고해서 반영해주시면 됩니다 :) 
cars에 대한 접근제어자도 붙여주세요~

sonypark commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,65 @@
+    public Cars(String userInput) {
+        validateNotNull(userInput);
+        validateHasComma(userInput);
+        addCar(userInput);
```
Cars는 Car를 리스트로 갖는 객체인데 생성자로 List<Car>로 받는게 자연스러울 것 같은데 어떻게 생각하시나요?

kiarakim replied:
네 저도 그 방식이 더 자연스럽다고 생각합니다.
입력 받는 InputView에서 String 으로 자동차 이름들을 받아오면 이를 검증해주는 InputValidation 클래스에 보내 List<Car>를 받아왔습니다. 받아온 리스트를 Car 생성자의 파라미터로 보내주니 Cars의 역할이 적절하게 분배된 것 같아요. 감사합니다.

sonypark commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,65 @@
+            // car.addDistance(new RandomNumberMaker().makeNumber());
+
+        }*/
+    }
```
randomNumbers를 받아서 배열의 인덱스로 자동차의 위치를 변경해주고 있네요.
randomNumbers의 인덱스가 변경돼서 들어오면 어떻게 되나요? 자동차와 랜덤값의 인덱스가 항상 같다는걸 보장할 수 있을까요?

kiarakim replied:
인덱스가 변경될거라는 가능성은 염두해두지 않았던 것 같아요.
자동차와 랜덤값의 인덱스가 항상 같다는 보장을 저는 아직 할 수 없어서 배열을 없앴습니다.
```
public void addDistance(int randomNumber, int index) {
	getCar(index).addDistance(randomNumber);
}
```
여기에 들어오는 randomNumber는 0부터 9 사이의 숫자인데 테스트 할 때 0부터 9 각각의 경우에 대한 값을 가정함으로써 랜덤값 테스트가 해결되는건지 궁금합니다. 
그것과 더불어 int index를 파라미터로 받아와 Car 리스트에 접근하고 있는데 이런 경우는 테스트가 어렵다고 할 수 있나요?? 테스트케이스 작성법을 공부하고 있지만 아직 명확한 기준을 세우지 못해 질문드립니다..

sonypark replied:
메서드는 addDistance 보다 move라는 네이밍은 어떠신가요?
- 메서드 인자로 randomNumber,index가 들어올때 어떤 값이 들어올지는 알 수 없습니다. 지금 구조에서는 int 타입만 있을뿐 어떤 값이 들어올 수 있는 상태입니다. 따라서 특정 값 이상일때 움직이고 이하일때 움직이지 않는다는 도메인 규칙에 따라 움직임을 보장할 수 없습니다.
- 그렇다면 도메인 규칙에 따라 전진할지 정지할지에 대한 여부를 반환하는 로직을 만들어 주입해주어야 합니다.
- move 메서드는 그 규칙을 인자로 받아 true면 전진 false면 정지하면 됩니다.
- Car의 move 메서드는 도메인 규칙에 따라 움직일지말지만 결정한다. 도메인 규칙은 외부에서 주입받는다.

`hint. interface, 전략패턴(Strategy pattern)`

sonypark commented on src/main/java/domain/Trial.java:
```diff
@@ -0,0 +1,31 @@
+
+    private final int trial;
+    private final static String PREFIX = "[0-9]";
+    private final String ERROR_TRIAL = "자연수로만 입력해주세요.";
```
여기도 컨벤션..!!

kiarakim replied:
넵...! 컨벤션 별표 백만개><!!

sonypark commented on src/main/java/domain/Trial.java:
```diff
@@ -0,0 +1,31 @@
+        validateNumber(input);
+        validateExtreme(input);
+        this.trial = Integer.parseInt(input);
+    }
```
횟수는 숫자만 들어올수있게 애초에 타입으로 강제하는건 어떨까요?

kiarakim replied:
네. int로 불러올 수 있도록 고친 후 입력을 받는 클래스(InputView)에서 String으로 받아와 객체 생성하는 부분을 int로 변경해주었습니다.
try-catch문을 사용해 정수로 받되 입력이 숫자가 아닌 경우 NumberFormatException을 날려주었는데 이 외에도 너무 큰 숫자를 입력했을 경우를 대비해 IllegalArgumentException도 날려주기로 했습니다.

여기서 궁금한 점이 있는데요, 제가 예외를 던질 때 특별한 이유가 아닌 이상 습관적으로 IllegalArgumentException을 사용합니다.
`적합하지 않거나(illegal) 적절하지 못한(inappropriate) 인자를 메소드에 넘겨주었을 때 발생합니다.`
라는 것을 보아 이 상황에서 해당 예외처리를 하는게 맞는지 여쭙고 싶습니다.

sonypark replied:
IllegalArgumentException 은 매개변수에 의해 의도치 않은 상황이 발생했을 때 사용하는 예외입니다. 이미 자바에서 구현해놓은 예외 클래스이고 상황에 맞게 잘 사용하신 것 같습니다. 서비스 예외에 맞게 커스텀해서 사용하기도 하지만 현재 코드에서는 IllegalArgumentException을 사용해도 충분할 것 같습니다.

sonypark commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,16 @@
+    public String askTrial(Scanner scanner){
+        System.out.println("시도할 횟수는 몇회인가요?");
+        return scanner.nextLine();
+    }
```
입력값에 대한 검증로직을 View에서 처리하는건 어떨까요?

kiarakim replied:
넵 감사합니다.
View는 입력에 관한 것만 처리하면 되는 줄 알았는데 피드백을 받아보니 클래스를 만들었으면 그에 맞는 적절한 역할을 주어야 한다는 것을 알게 되었습니다.
이에 피드백을 반영해 View에 검증로직을 만들어 주었는데 혹시 InputView 라는 클래스에선 Scanner를 통한 입력만을 담당하고 따로 InputValidation 이라는 클래스를 만들어 여기에 검증로직을 작성하는 것은 비효율적인 방법인지 궁금합니다

sonypark replied:
InputValidation을 위한 클래스를 따로 만들어도 괜찮습니다. 그런데 클래스로 분리할만큼의 역할을 하고 그만한 가치가 있는지는 생각해봐야할 것 같습니다. 해당 클래스가 InputView 이외에 다른 곳에서 쓰일 일이 있을지, InputView 내부에 검증 코드를 포함했을때와 비교해 어떤 장점이 있을지. 이러한 것을 고려하여 키아라가 판단해서 구현해주시면 될 것 같아요!

sonypark commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,24 @@
+    void checkLength(String input) {
+        assertThatThrownBy(() ->
+                new Car(input)
+        ).isInstanceOf(IllegalArgumentException.class);
```
어떤 테스트인지 구체적으로 명시해주는게 좋을 것 같아요.
`@DisplayName()` 을 사용해보세요

kiarakim replied:
넵! 메소드 이름만 구체적으로 쓰면 될 줄 알았는데 역시 DisplayName을 명시해주어야 테스트 돌릴 때 파악하기 좋겠네요.

sonypark commented on src/test/java/domain/CarsTest.java:
```diff
@@ -0,0 +1,59 @@
+        assertThatThrownBy(() ->
+                new Cars("aa,bb,cc")
+        ).isNotInstanceOf(IllegalArgumentException.class);
+    }
```
가장 중요한 move에 대한 테스트가 없네요🥲

kiarakim replied:
move에 대한 테스트라는 것은 변수 distance의 값이 증가한다는 것을 테스트하는건가요? 아니면 그에 따라 '-'로 표현되는 거리가 찍히는 테스트라는 건가요?

코드 리펙터링을 하면서 몇몇 코드가 다른 클래스로 넘어가서 move()에 대한 테스트는 CarTest에 최대한 많이 작성해보았습니다...

테스트 코드는 넘 어려운 것 같아요..ㅠ

sonypark replied:
move 메서드에 대한 코멘트를 참고하셔서 테스트를 작성해보세요🙌🏻
이번 미션에서는 move에 대한 로직 구현과 테스트가 가장 중요하다고 할 수 있어요.
가장 어려운 부분이자 중요한 부분이니 고민할 가치가 충분히 있다고 생각합니다 :)

sonypark reviewed (changes requested):
키아라 안녕하세요. 리뷰어 소니입니다! 반가워요 🙂
첫 미션은 어떠셨나요? 어려운 부분이 있었다면 어떤게 어려웠을지도 궁금합니다 :)
코멘트에도 남겼지만 아래 사항을 우선순위로 두고 먼저 개선해주세요~
1. 자바 코드 컨벤션 -> 코드컨벤션은 꼭 지켜주세요. 프리코스에서도 컨벤션을 강조했을텐데 한 번 더 확인해주세요.
2. 메서드는 하나의 역할을 하도록 분리
3. 단위 테스트 코드 -> 테스트도 가독성이 중요합니다. 어떤 테스트인지 설명을 달아주세요 :)

kiarakim committed:
feat : Scanner 기능을 Controller에서 InputView로 전환

kiarakim committed:
refactor : static과 final 제어자 순서 변경, 매직넘버 제거, 캐스팅 오류 해결

kiarakim committed:
refactor : 매직넘버 제거

kiarakim committed:
refactor : 코드 컨벤션에 따른 제어자 순서 변경

kiarakim committed:
refactor : static 제어자 추가

kiarakim committed:
refactor : 자동차 순위 판단 로직을 Car -> Winner 클래스로 이동

Winner 클래스에서 Collection.max()를 사용하기 위해 Car 클래스에서 Complarable<Car> 을 implement 해주고 compareTo() 메서드를 오버라이딩해준 것을 삭제 후 Winner 클래스에서 순위 판별 로직 구현

kiarakim committed:
refactor : Cars 생성자로 List<Car> 받게 변경

kiarakim committed:
refactor : 주석 제거

kiarakim committed:
refactor : 주석 제거

kiarakim committed:
refactor : 랜덤값을 리스트가 아닌 정수값으로 불러와 검증

kiarakim committed:
style : CheckStyle 적용

kiarakim committed:
test : Car 테스트 코드 작성

kiarakim committed:
test : Cars 테스트 코드 작성

kiarakim committed:
test : Trial 테스트 코드 작성

kiarakim committed:
test : Winner 테스트 코드 작성

sonypark reviewed (approved):
키아라 안녕하세요. 의견 남겨주신것 잘봤습니다.
의도한 부분을 이해해서 잘 적용해주셨네요. 2단계로 진행해주시면 될 것 같습니다.
2단계도 화이팅!
